// @manuth/package-json-editor/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JSONObjectBase = exports.JSONObject = exports.Package = exports.Person = exports.OrderedDependencyCollection = exports.DependencyCollection = exports.BugInfo = exports.LoadLogic = exports.GenerationLogic = exports.PropertyDictionary = exports.List = exports.Dictionary = exports.Collection = exports.AlphabeticalList = exports.AlphabeticalDictionary = void 0;
var AlphabeticalDictionary_1 = require("./Collections/AlphabeticalDictionary");
Object.defineProperty(exports, "AlphabeticalDictionary", { enumerable: true, get: function () { return AlphabeticalDictionary_1.AlphabeticalDictionary; } });
var AlphabeticalList_1 = require("./Collections/AlphabeticalList");
Object.defineProperty(exports, "AlphabeticalList", { enumerable: true, get: function () { return AlphabeticalList_1.AlphabeticalList; } });
var Collection_1 = require("./Collections/Collection");
Object.defineProperty(exports, "Collection", { enumerable: true, get: function () { return Collection_1.Collection; } });
var Dictionary_1 = require("./Collections/Dictionary");
Object.defineProperty(exports, "Dictionary", { enumerable: true, get: function () { return Dictionary_1.Dictionary; } });
var List_1 = require("./Collections/List");
Object.defineProperty(exports, "List", { enumerable: true, get: function () { return List_1.List; } });
var PropertyDictionary_1 = require("./Collections/PropertyDictionary");
Object.defineProperty(exports, "PropertyDictionary", { enumerable: true, get: function () { return PropertyDictionary_1.PropertyDictionary; } });
var GenerationLogic_1 = require("./GenerationLogic");
Object.defineProperty(exports, "GenerationLogic", { enumerable: true, get: function () { return GenerationLogic_1.GenerationLogic; } });
var LoadLogic_1 = require("./LoadLogic");
Object.defineProperty(exports, "LoadLogic", { enumerable: true, get: function () { return LoadLogic_1.LoadLogic; } });
var BugInfo_1 = require("./Management/BugInfo");
Object.defineProperty(exports, "BugInfo", { enumerable: true, get: function () { return BugInfo_1.BugInfo; } });
var DependencyCollection_1 = require("./Management/DependencyCollection");
Object.defineProperty(exports, "DependencyCollection", { enumerable: true, get: function () { return DependencyCollection_1.DependencyCollection; } });
var OrderedDependencyCollection_1 = require("./Management/OrderedDependencyCollection");
Object.defineProperty(exports, "OrderedDependencyCollection", { enumerable: true, get: function () { return OrderedDependencyCollection_1.OrderedDependencyCollection; } });
var Person_1 = require("./Management/Person");
Object.defineProperty(exports, "Person", { enumerable: true, get: function () { return Person_1.Person; } });
var Package_1 = require("./Package");
Object.defineProperty(exports, "Package", { enumerable: true, get: function () { return Package_1.Package; } });
var JSONObject_1 = require("./Utilities/JSONObject");
Object.defineProperty(exports, "JSONObject", { enumerable: true, get: function () { return JSONObject_1.JSONObject; } });
var JSONObjectBase_1 = require("./Utilities/JSONObjectBase");
Object.defineProperty(exports, "JSONObjectBase", { enumerable: true, get: function () { return JSONObjectBase_1.JSONObjectBase; } });
