// @manuth/package-json-editor/lib/Management/Person.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Person = void 0;
const ParseAuthor = require("parse-author");
const StringifyAuthor = require("stringify-author");
const JSONObjectBase_1 = require("../Utilities/JSONObjectBase");
/**
 * Represents a person.
 */
class Person extends JSONObjectBase_1.JSONObjectBase {
    /**
     * Initializes a new instance of the `Person` class.
     *
     * @param person
     * A variable containing info about the person.
     */
    constructor(person) {
        var _a, _b, _c;
        super();
        let info;
        if (typeof person === "string") {
            info = ParseAuthor(person);
        }
        else {
            info = person;
        }
        this.Name = (_a = info.name) !== null && _a !== void 0 ? _a : null;
        this.EMail = (_b = info.email) !== null && _b !== void 0 ? _b : null;
        this.URL = (_c = info.url) !== null && _c !== void 0 ? _c : null;
    }
    /**
     * Gets a json-object representing the person.
     *
     * @returns
     * A json-object representing the person.
     */
    ToJSON() {
        let result = StringifyAuthor({
            name: this.Name,
            email: this.EMail,
            url: this.URL
        });
        if (result.length === 0) {
            return null;
        }
        else {
            return result;
        }
    }
}
exports.Person = Person;
